script: Implement Compression API

kkoyung opened the pull request:
The patch implements Compression (https://compression.spec.whatwg.org/) with the compression and decompression provided by the `flate2` crate (https://crates.io/crates/flate2).

`flate2` supports several different backends, controlled through the crate's features. By default, it uses `miniz_oxide` (https://crates.io/crates/miniz_oxide).

`flate2` provides three modules `read`, `write` and `bufread` which work on instances of the `std::io::Read`, `std::io::Write` and `std::io::Bufread` traits, respectively. The `write` module is chosen in the patch since it matches the streaming model in the specification.

Testing: Enable WPT for Compression API, and introduce WPT expectation.

kkoyung committed:
Implement Compression API

https://compression.spec.whatwg.org/#decompressionstream

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Introduce WPT expectation for Compression API

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Add spec link top tidy.py

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Update WPT expectation

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Enable WPT tests for Compression API

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Add CompressionStream and DecompressionStream to mozilla/interfaces.https.html

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Fix spec link

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Clear buffer to save memory

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Explain more in comment note

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

TimvdLippe commented on components/script/dom/compressionstream.rs:
```diff
@@ -0,0 +1,276 @@
+        format: CompressionFormat,
+        can_gc: CanGc,
+    ) -> Fallible<DomRoot<CompressionStream>> {
+        // Step 1. If format is unsupported in CompressionStream, then throw a TypeError.
```
Nit: these spec comments usually refer to the `Constructor`. We typically implement that in the `Constructor` and keep `new_with_proto` small.

TimvdLippe commented on components/script/dom/compressionstream.rs:
```diff
@@ -0,0 +1,276 @@
+
+    /// <https://streams.spec.whatwg.org/#dom-generictransformstream-readable>
+    fn Readable(&self) -> DomRoot<ReadableStream> {
+        self.transform.get_readable()
```
Nit: add a spec comment for these as well:

> The readable getter steps are to return this’s transform.[[readable]].

TimvdLippe commented on tests/wpt/meta/compression/compression-bad-chunks.tentative.any.js.ini:
```diff
@@ -0,0 +1,62 @@
+  expected: ERROR
+
+[compression-bad-chunks.tentative.any.worker.html]
+  [chunk of type SharedArrayBuffer should error the stream for gzip]
```
Above I see this implemented, yet the tests are failing. Do you know what we are missing here?

jdm replied:
Correct.

TimvdLippe commented on components/script/dom/decompressionstream.rs:
```diff
@@ -0,0 +1,305 @@
+    format: CompressionFormat,
+
+    // <https://compression.spec.whatwg.org/#decompressionstream-context>
+    #[ignore_malloc_size_of = "defined in flate2"]
```
Can we define an external malloc size implementation for this?

TimvdLippe reviewed (approved):
Did a quick review. Given that almost all WPT tests pass, I think it is pretty much complete. Only one remaining question regarding the bad chunks.

kkoyung committed:
Move steps from `new_with_proto`` to `Constructor`

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Add spec text to `readable`/`writable` getters

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Prettier `use` statements

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

TimvdLippe commented on components/script/dom/compressionstream.rs:
```diff
@@ -0,0 +1,277 @@
+        format: CompressionFormat,
+        can_gc: CanGc,
+    ) -> Fallible<DomRoot<CompressionStream>> {
+        Ok(reflect_dom_object_with_proto(
```
Nit: afaik this can never fail and we typically return `DomRoot<T>` from `new_with_proto`. Therefore, we can remove the `Fallible` and `Ok` here

TimvdLippe reviewed (approved):
Last nit, but this still LGTM

kkoyung committed:
Infallible `new_with_proto`

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

Taym95 commented on components/script/dom/compressionstream.rs:
```diff
@@ -0,0 +1,277 @@
+    buffer_source.safe_to_jsval(cx, rval.handle_mut());
+    controller.enqueue(cx, global, rval.handle(), can_gc)?;
+
+    Ok(())
```
You already clear the compressor’s buffer after each transformed chunk to save memory you probably shoudl Do the same in the flush path to avoid retaining the final output slice?!

Taym95 commented on components/script/dom/decompressionstream.rs:
```diff
@@ -0,0 +1,303 @@
+    // NOTE: In our implementation, the enum type of context already indicates the format.
+    let mut decompressor = ds.context.borrow_mut();
+    let offset = decompressor.get_ref().len();
+    let is_ended = decompressor
```
you compute is_ended before calling try_finish, but you enqueue before checking !is_ended. That can emit bytes and then throw?!

kkoyung replied:
I also think it is weird, but the spec does emit bytes no matter whether the test passes or not.
https://compression.spec.whatwg.org/#decompress-flush-and-enqueue

p.s. I made a little change in the new commit. I flush and copy the buffer before doing the test. This guarantees the extra bytes won't go into the result. However, it does not resolve the weirdness here.

Taym95 commented on components/script/dom/compressionstream.rs:
```diff
@@ -0,0 +1,277 @@
+    let offset = compressor.get_ref().len();
+    compressor
+        .write_all(&chunk)
+        .map_err(|_| Error::Type("Fail to write all to compressor".to_string()))?;
```
```suggestion
        ..map_err(|_| Error::Type("CompressionStream: write_all() failed".to_string()))?;
```

Taym95 replied:
Apply similarly for flush, try_finish, and the decompressor paths, this is  helpful

Taym95 commented on components/script/dom/decompressionstream.rs:
```diff
@@ -0,0 +1,303 @@
+        // Step 2. Set this’s format to format.
+        // Step 5. Set this’s transform to a new TransformStream.
+        let transform = TransformStream::new_with_proto(global, None, can_gc);
+        let this = DecompressionStream::new_with_proto(global, proto, &transform, format, can_gc);
```
I would avoid using `this` and `that` in naming, `Set this’s` means to set current DecompressionStream's format

Taym95 commented on components/script/dom/compressionstream.rs:
```diff
@@ -0,0 +1,277 @@
+) -> Fallible<()> {
+    // Step 1. If chunk is not a BufferSource type, then throw a TypeError.
+    let conversion_result =
+        ArrayBufferViewOrArrayBuffer::safe_from_jsval(cx, chunk, ()).map_err(|_| {
```
this code is being copied and past lot, can we move it to hepler ?

kkoyung replied:
I think `BufferSource` is the good place to implement the helper, but I saw this discussion (https://github.com/servo/servo/pull/38112#discussion_r2218143213) which tried to implement it before, and it fails. Look like we need some investigation on it.

Taym95 commented on components/script/dom/transformstreamdefaultcontroller.rs:
```diff
@@ -436,6 +519,54 @@ impl TransformStreamDefaultController {
+                        // Step 6.1 If this throws an exception e,
+                        let realm = enter_realm(self);
+                        let p = Promise::new_in_current_realm((&realm).into(), can_gc);
+                        // return a promise rejected with e.
```
you can use `Promise::new_rejected` here?

kkoyung replied:
`Promise::new_rejected` takes `value: impl SafeToJSValConvertible`, but `e` is a `dom::bindings::error::Error`, which currently does not implement this trait. If we use `Promise::new_rejected`, we still need extra few lines to do the conversion.

kkoyung committed:
Avoid using `this` as variable name

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Better error messages

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Also clear inner buffer in flush path

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Flush and copy bytes before doing test

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Revert "Flush and copy bytes before doing test"

This reverts commit 3da72d3ad38a9dc24180a6fc0b3041731cb71a86.

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Helper to convert chunk to vec

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Add expectation of new WPT tests from upstream

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Fix timing of throwing error when decompressing input with extra pad

This helps passing WPT tests from
`compression/decompression-extra-input.any.js`.

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung commented:
Rebased to resolve merge conflict. Also appended a fix for passing new WPT tests from upstream.

kkoyung committed:
Formatting

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Pin flate2 to 1.1, instead of 1.1.2

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung commented on Cargo.toml:
```diff
@@ -61,6 +61,7 @@ embedder_traits = { path = "components/shared/embedder" }
 encoding_rs = "0.8"
 env_logger = "0.11"
 euclid = "0.22"
+flate2 = "1.1.2"
```
Thanks for pointing out. It's now pinned to 1.1.

TimvdLippe commented:
This PR still LGTM. @Taym95 do you have any further review feedback or does somebody else wants to take a look? If not, then I would opt for merging this PR.

jdm commented:
Let's press the button!

kkoyung committed:
Add CanGc to safe_to_jsval

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Update WPT expectation in response to change in test name in #40172

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung commented:
Rebased with 2 new commits to resolve the merge error

kkoyung committed:
Add CompressionStream and DecompressionStream to interfaces.worker.js

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Update MANIFEST

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>
